nengo/learning_rules.py

```diff
@@ -20,7 +23,7 @@ def coerce(self, instance, size_in):
                 instance, size_in)  # IntParam validation
 
 
-class LearningRuleType(FrozenObject):
```

Comment:
So the idea behind FrozenObjects is that they're not supposed to support Defaults. I remember the idea behind this was so that FrozenObjects could be used in multiple places (e.g. the same learning rule type on multiple connections) or copied without issue. But I can't remember the exact details as to why having Defaults and being Frozen could not go together.

Part of the reason might be because FrozenObjects themselves are often used as defaults for other things. For example, the default for ``Ensemble.neuron_type`` is ``LIF()``. If ``LIF`` had its own defaults for ``tau_rc`` and ``tau_ref``, then changing ``model.config[nengo.LIF].tau_rc`` could change the default tau_rc for an ensemble, but also change that default other places, too. Things could get complicated or have unintended consequences.

For learning rules, that's less of an issue, since they're currently not used as defaults anywhere that I know of. My main point is just that I had designed FrozenObject to be separate from the Default system, so if we're bringing them together sometimes, I just want to make sure that that makes sense and doesn't cause any unforeseen problems.

Reply:
That's the reason all the parameters are marked as `readonly`, so that they can be used with `FrozenObject` (this is explicitly checked for in [FrozenObject](https://github.com/nengo/nengo/blob/master/nengo/params.py#L498), so I think it is part of the intended/expected behaviour).